pipeline/lib/json_stream.ml

```diff
@@ -38,30 +37,64 @@ let json_step stack chr =
   | _, (('{' | '[' | '"') as chr) -> chr :: stack
   | _ -> stack
 
-let make_json_parser () = { current = Buffer.create 16; stack = [] }
+let make_json_parser () =
+  { current = Buffer.create 16; stack = [ '\n' ]; lines = 1L }
+
+let count_newlines str =
+  let ( ++ ) = Int64.add in
+  let rec position str start count =
+    try
+      let start = String.index_from str start '\n' in
+      position str (start + 1) (count ++ 1L)
+    with Not_found -> count
+  in
+  position str 0 0L
 
 let json_step state chr =
+  let state =
+    match chr with
+    | '\n' -> { state with lines = Int64.add state.lines 1L }
+    | _ -> state
+  in
   match json_step state.stack chr with
   | [] ->
-      if Buffer.length state.current = 0
-      then (None, { state with stack = [] })
+      if Buffer.length state.current = 0 then
+        (None, None, { state with stack = [] })
       else (
         Buffer.add_char state.current chr;
         let str = Buffer.contents state.current in
-        (Some str, make_json_parser ()))
+        let st = make_json_parser () in
+        let ( + ) = Int64.add in
+        (Some str, Some (1L + state.lines), { st with lines = state.lines }))
   | hd :: _ as stack ->
-      if hd <> '\n' then Buffer.add_char state.current chr;
-      (None, { state with stack })
+      if chr <> '\n' || hd = '"' then Buffer.add_char state.current chr;
+      let pos =
+        if Buffer.length state.current == 1 then Some state.lines else None
+      in
+      (None, pos, { state with stack })
 
-let json_steps (parsed, state) str =
+let json_steps (parsed, positions, state) str =
   String.fold_left
-    (fun (parsed, state) chr ->
-      let opt_json, state = json_step state chr in
-      let parsed =
-        match opt_json with Some json -> json :: parsed | None -> parsed
+    (fun (parsed, positions, state) chr ->
+      let opt_json, pos, state = json_step state chr in
+      let parsed, positions =
+        match (opt_json, pos) with
+        | Some json, Some _end -> (
+            match positions with
+            | (start, _) :: positions ->
+                (json :: parsed, (start, pos) :: positions)
+            | [] -> (json :: parsed, positions))
+        | None, Some _start -> (
+            match positions with
+            | [] -> (parsed, (pos, None) :: positions)
+            | ((Some _, Some _) as hd) :: positions ->
+                (parsed, (pos, None) :: hd :: positions)
+            | (None, Some y) :: positions -> (parsed, (pos, Some y) :: positions)
+            | _ :: positions -> (parsed, (pos, None) :: positions))
```

Comment:
The code could be clearer if the `positions` were stored in the same list as `parsed` as they are supposed to have the same length in the end... but it's harder to maintain this invariant when you have `string list * range list` rather than `(string * range) list`.

I would add a `start_line : int` to the `state` to remember when we first started adding stuff to the buffer, so that we can push `(state.start_line, state.lines, str_json)` to the `parsed` list when we are done. This would also avoid the pattern matching on `positions` to discover the current range start/stop :)